cli/src/commands/cvms/logs-handler.ts

```diff
@@ -0,0 +1,131 @@
+import type { CommandContext } from "@/src/core/types";
+import { logger, setJsonMode } from "@/src/utils/logger";
+
+export interface LogsHandlerConfig<TOptions> {
+	logType: "container" | "serial";
+	fetchLogs: (appId: string, options: TOptions) => Promise<string>;
+	streamLogs: (
+		appId: string,
+		onData: (data: string) => void,
+		options: TOptions,
+		signal?: AbortSignal,
+	) => Promise<void>;
+}
+
+export interface BaseLogsInput {
+	json?: boolean;
+	follow?: boolean;
+	tail?: number;
+	timestamps?: boolean;
+}
+
+/**
+ * Extract CVM ID from context
+ */
+export function extractCvmId(context: CommandContext): string | null {
+	if (!context.cvmId) return null;
+	return (
+		context.cvmId.id ||
+		context.cvmId.app_id ||
+		context.cvmId.uuid ||
+		context.cvmId.name ||
+		null
+	);
+}
+
+/**
+ * Create a logs command handler with shared logic
+ */
+export function createLogsHandler<TInput extends BaseLogsInput, TOptions>(
+	config: LogsHandlerConfig<TOptions>,
+	buildOptions: (input: TInput) => TOptions,
+) {
+	return async function handler(
+		input: TInput,
+		context: CommandContext,
+	): Promise<number> {
+		setJsonMode(input.json);
+
+		if (!context.cvmId) {
+			context.fail("No CVM ID provided. Please specify a CVM ID.");
+			return 1;
+		}
+
+		const appId = extractCvmId(context);
+		if (!appId) {
+			context.fail("Could not determine CVM identifier");
+			return 1;
+		}
+
+		const options = buildOptions(input);
+		const logTypeName = config.logType === "serial" ? "serial" : "container";
+
+		try {
+			if (input.follow) {
+				if (input.json) {
+					context.fail("Cannot use --json with --follow");
+					return 1;
+				}
+
+				logger.info(`Streaming ${logTypeName} logs (press Ctrl+C to stop)...`);
+				logger.break();
+
+				const abortController = new AbortController();
+				const cleanup = () => {
+					abortController.abort();
+					logger.break();
+					logger.info("Stopped streaming logs");
+					process.exit(0);
+				};
+
+				process.on("SIGINT", cleanup);
+				process.on("SIGTERM", cleanup);
+
+				try {
+					await config.streamLogs(
+						appId,
+						(data) => process.stdout.write(data),
+						options,
+						abortController.signal,
+					);
+				} catch (error) {
+					if ((error as Error).name === "AbortError") {
+						return 0;
+					}
+					throw error;
+				} finally {
+					process.removeListener("SIGINT", cleanup);
+					process.removeListener("SIGTERM", cleanup);
+				}
+
+				return 0;
+			}
+
+			// Non-streaming mode
+			const spinner = logger.startSpinner(`Fetching ${logTypeName} logs`);
```

Comment:
Even in non-streaming mode, we should remove the spinner entirely. This will make the pipeline completely unusable.

e.g. `phala logs xx | grep joking`